feat: implement product card and picking logic

nsilva-spryker opened the pull request:
-Add new product card to pick items
-Adapte logic on list to handle each event of edit, submit and change items

Closes [HRZ-2411](https://spryker.atlassian.net/browse/HRZ-2411)


[HRZ-2411]: https://spryker.atlassian.net/browse/HRZ-2411?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

nsilva-spryker committed:
feature: new pr after merge issues with development and previous branches

nsilva-spryker committed:
feature: add alias to minus and add icons in fulfillment app

nsilva-spryker committed:
feature: refactoring some code

Sinucid commented on libs/domain/picking/src/components/picking-product-card/picking-product-card.component.ts:
```diff
@@ -0,0 +1,262 @@
+} from '../../models';
+import { styles } from './picking-product-card.styles';
+
+export const EVENT_CHANGE_NUMBER_OF_PICKED = 'oryx.change-number-of-picked';
```
better to keep aliases for events inside model

Sinucid commented on libs/domain/picking/src/components/picking-product-card/picking-product-card.component.ts:
```diff
@@ -0,0 +1,262 @@
+export class PickingProductCardComponent extends LitElement {
+  static styles = styles;
+
+  @property() productItem?: PickingListItem;
```
I believe that we should implement an interface for this public properties

nsilva-spryker replied:
What are the benefits of having an interface for just this two properties?

Sinucid replied:
This is public interface for the element. This allows us:
1) To get available properties and their types when query the element
2) Use this interface for storybook as a type for storybook properties

But you might be right. If there are no stories for this component, this can be useless. But if it requires stories - better to create interface and reuse it. Up to you

nsilva-spryker replied:
Let's change/adapt this if we need stories for that in another sub-task or independent task

Sinucid commented on libs/domain/picking/src/components/picking-product-card/picking-product-card.component.ts:
```diff
@@ -0,0 +1,262 @@
+              <span>${this.productItem?.orderItem.sku}</span>
+            </div>
+
+            <img
```
you can use oryx-image. It also supports fallback image

Sinucid commented on libs/domain/picking/src/components/picking-product-card/picking-product-card.component.ts:
```diff
@@ -0,0 +1,262 @@
+
+  @state() isCorrectNumberOfPickedProvided?: boolean = true;
+
+  protected currentNumberOfPicked$ = new BehaviorSubject(0);
```
Why are the observables? Looks like just lit states is enough here

Sinucid commented on libs/domain/picking/src/components/picking-product-card/picking-product-card.component.ts:
```diff
@@ -0,0 +1,262 @@
+  }
+
+  protected confirmPartialPicking(): void {
+    this.dispatchPickingEvents(EVENT_CHANGE_NUMBER_OF_PICKED, {
```
I believe that 2 separate events dispatching are overhead. Can submit also provides a number of picked in details?

Sinucid commented on libs/domain/picking/src/components/picking-product-card/picking-product-card.component.ts:
```diff
@@ -0,0 +1,262 @@
+  }
+
+  protected renderEditStatus(): TemplateResult {
+    this.summaryInfo = this.getSummaryInfo();
```
Why do you need to store it to the property? This doesn't have sense

Sinucid replied:
```
protected renderEditStatus(): TemplateResult {
    let text = '';
    let label = '';
    let subtext = '';
    
    if (this.status === ItemsFilters.Picked) {
        text = `${this.productItem.numberOfPicked}/${this.productItem.quantity}`;

        if (this.productItem.numberOfPicked < this.productItem.quantity) {
            label = i18n('picking.product-card.items-picked');
        } else {
            subtext = i18n('picking.product-card.all-items-picked')
        }
    } else if (this.status === ItemsFilters.NotFound) {
        text = `${this.productItem.numberOfNotPicked}/${this.productItem.quantity}`;

        if (this.productItem.numberOfNotPicked < this.productItem.quantity) {
            label = i18n('picking.product-card.items-not-found');
        } else {
            subtext = i18n('picking.product-card.no-items-found')
        }
    }

    return html` <div class="summary-info">
        <p>${text} ${label}</p>
        ${when(
            subtext,
            () => <p>${subtext}</p>
        )}
      </div>
      <oryx-button>
        <button @click=${this.editProductPicking}>
          <oryx-icon type="edit"></oryx-icon>
          ${i18n('picking.product-card.edit-items')}
        </button>
      </oryx-button>`;
  }
```

nsilva-spryker replied:
It make sense to separated into getSummaryInfo() method, by doing that the code are more modular and easier to understand

Sinucid replied:
I don't think so. This has sense if you are reusing this part somewhere or its to long. My version is contains ~30 rows that is not so long and fits one page in editor. Not a case I would say

nsilva-spryker replied:
Your point is also right, but from my point of view, extracting this logic to another method and keep 'renderEditStatus' more readable and more focus on the html would be more clean.
Nevertheless I would adapt the code to your version

Sinucid commented on libs/domain/picking/src/components/picking-product-card/picking-product-card.component.ts:
```diff
@@ -0,0 +1,262 @@
+        this.productItem,
+        () => html`
+          <oryx-card>
+            <div slot="heading" class="title">
```
can we use oryx-heading here?

Sinucid commented on libs/domain/picking/src/components/picking-product-card/picking-product-card.component.ts:
```diff
@@ -0,0 +1,262 @@
+            <div slot="heading" class="title">
+              ${this.productItem?.orderItem?.name}
+            </div>
+            <div class="subtitle">
```
Why do you need to put span inside div?

Sinucid commented on libs/domain/picking/src/components/picking-product-card/picking-product-card.component.ts:
```diff
@@ -0,0 +1,262 @@
+                    @submit=${this.onChangeQuantity}
+                  ></oryx-cart-quantity-input>
+
+                  <div class="edit-quantity-info">
```
The same, span inside div

Sinucid commented on libs/domain/picking/src/components/picking-product-card/picking-product-card.component.ts:
```diff
@@ -0,0 +1,262 @@
+
+                  <oryx-button>
+                    <button
+                      type="submit"
```
not needed, button without type works as submit

Sinucid commented on libs/domain/picking/src/components/picking-product-card/picking-product-card.styles.ts:
```diff
@@ -0,0 +1,84 @@
+    object-fit: cover;
+    max-width: 182px;
+
+    &__fade {
```
left over?

Sinucid commented on libs/domain/picking/src/components/picking/picking.component.ts:
```diff
@@ -3,26 +3,41 @@ import { asyncState, i18n, observe, valueType } from '@spryker-oryx/utilities';
+
+  protected pickingUpdate$ = new BehaviorSubject<string | null>(null);
+
+  protected tabs$ = combineLatest([
```
Why do you need to combine them?

Sinucid commented on libs/domain/picking/src/components/picking/picking.component.ts:
```diff
@@ -53,6 +68,46 @@ export class PickingComponent extends LitElement {
+    const detail = (event as CustomEvent<ProductItemPickedEvent>).detail;
+
+    this.pickingList?.items.map((item) => {
+      if (
```
this is working not properly, this changing elements of reactive array doesn't trigger update. You need to ensure whole approach

Sinucid commented on libs/domain/picking/src/components/picking/picking.component.ts:
```diff
@@ -53,6 +68,46 @@ export class PickingComponent extends LitElement {
+
+    this.pickingList?.items.map((item) => {
+      if (
+        item.product.id === detail.productId &&
```
item.product.id === detail.productId && detail.numberOfPicked && detail.numberOfPicked <= item.quantity

nsilva-spryker committed:
Merge branch 'development' into feature/hrz-2411-implement-picking-logic

# Conflicts:
#	libs/domain/picking/src/components/picking/picking.component.ts
#	libs/domain/picking/src/routes.ts

nsilva-spryker committed:
feature: apply some cr suggestions

nsilva-spryker committed:
feature: update events and how reactive the tabs list are updating

nsilva-spryker committed:
feature: refactor some code

nsilva-spryker committed:
feature: refactor some code, fix index key error

nsilva-spryker committed:
Merge branch 'development' into feature/hrz-2411-implement-picking-logic

nsilva-spryker committed:
feat: set placeholder when all products are picked

- feature: add success placeholder when all product picking are done
 -feature: refactor some code

nsilva-spryker committed:
feature: set modal component in picking list to not render multiple models for every product

nsilva-spryker committed:
feature: refactor some code

Sinucid commented on libs/domain/picking/src/components/picking-product-card/picking-product-card.component.ts:
```diff
@@ -0,0 +1,184 @@
+    this.currentNumberOfPicked =
+      this.currentNumberOfPicked ?? this.productItem?.numberOfPicked;
+
+    if (this.productItem?.product.id) {
```
```
this.dispatchPickingEvents(EVENT_SUBMIT, {
        productId: this.productItem.product.id!,
        numberOfPicked: this.currentNumberOfPicked,
});
```

nsilva-spryker replied:
Optional chain expressions can return undefined by design - using a non-null assertion is unsafe and wrong

Sinucid replied:
Yes, but in render function you are rendering form with submit handling only when `this.productItem` is truthy, that means it can't dispatch submit till `this.productItem` is falsy.

nsilva-spryker replied:
I think for this case using non-assertion could be, like this:
`productId: this.productItem!.product.id`

The productItem should never be null

Sinucid commented on libs/domain/picking/src/components/picking-product-card/picking-product-card.component.ts:
```diff
@@ -0,0 +1,184 @@
+
+  protected editProductPicking(): void {
+    this.dispatchPickingEvents(EVENT_EDIT, {
+      productId: this.productItem?.product.id,
```
`productId: this.productItem.product.id!,`

Sinucid commented on libs/domain/picking/src/components/picking/picking.component.ts:
```diff
@@ -1,53 +1,142 @@
+  );
+
+  @state()
+  protected partialPicking: PartialPicking | undefined;
```
`protected partialPicking: PartialPicking | null = null;`

You can use this one as a trigger for confirmation modal. It will contain partial picking data or null. When it has data -> show modal. And you can clean up `isConfirmPickingDialogOpen`

Sinucid commented on libs/domain/picking/src/components/picking/picking.component.ts:
```diff
@@ -1,53 +1,142 @@
-    map((list) => [
+  static styles = styles;
+
+  protected isConfirmPickingDialogOpen$ = new BehaviorSubject(false);
```
Why do you need observable here? What the point? 

Why not just a lit state?

```
@state()
protected isConfirmPickingDialogOpen = false;
```

Sinucid commented on libs/domain/picking/src/components/picking/picking.component.ts:
```diff
@@ -56,29 +145,117 @@ export class PickingComponent extends PickingListMixin(LitElement) {
+  protected renderTabContents(tabs: PickingTab[]): TemplateResult {
+    return html`
+      ${when(
+        tabs.length,
```
It will always have length, this check is useless

Sinucid commented on libs/domain/picking/src/components/picking-product-card/picking-product-card.component.ts:
```diff
@@ -0,0 +1,184 @@
+            <oryx-image
+              .src="${this.productItem?.product?.image}"
+              alt="${ifDefined(this.productItem?.orderItem?.name)}"
+              class="${this.status === ItemsFilters.NotFound
```
use `classMap` lit directive please

Sinucid commented on libs/domain/picking/src/components/picking/picking.component.ts:
```diff
@@ -56,29 +145,117 @@ export class PickingComponent extends PickingListMixin(LitElement) {
+    `;
   }
 
   protected renderFallback(): TemplateResult {
```
```
protected renderFallback(): TemplateResult | void {
    if (!this.pickingList?.items.every(
      (item) => item.status === ItemsFilters.Picked
    )) {
      return html`
        <div class="picking-complete">
          <div class="img-wrap">
            <oryx-image resource="picking-items-processed"></oryx-image>
          </div>
          <h3 class="title-empty">${i18n(`picking.great-job`)}!</h3>
          <p>${i18n(`picking.all-items-are-processed`)}!</p>
        </div>`
    }
  }
```

nsilva-spryker replied:
We want the opposite if that, to know that every item from the picking list was picked already

Sinucid replied:
Ok, just change the value of the logical expression to the opposite:
`if (this.pickingList?.items.every(`

Sinucid commented on libs/domain/picking/src/components/picking/picking.component.ts:
```diff
@@ -56,29 +145,117 @@ export class PickingComponent extends PickingListMixin(LitElement) {
+        </div>
+
+        <span>
+          You only picked
```
Add this ToDo please:
`//ToDo: should be refactored when i18n is supporting elements as a parts of templates`

nsilva-spryker replied:
let's just use the `${i18n('picking.product-card.you-only-picked')}` for now

Sinucid commented on libs/domain/picking/src/components/picking/picking.component.ts:
```diff
@@ -56,29 +145,117 @@ export class PickingComponent extends PickingListMixin(LitElement) {
+    productIndex: number,
+    numberOfPicked?: number
+  ): void {
+    this.items[productIndex].numberOfPicked = numberOfPicked ?? 0;
```
Do items contain some unique identifier? Using index is not very safe

nsilva-spryker replied:
Since the items are not being push or pop, i don't think there are high risk on this. Also we are providing type annotations for the callback function, that ensures typeScript infers the correct type and avoid potential type or runtime errors

tarashaliuk commented on libs/domain/picking/src/components/picking-product-card/picking-product-card.component.ts:
```diff
@@ -0,0 +1,184 @@
+} from '../../models';
+import { styles } from './picking-product-card.styles';
+
+export class PickingProductCardComponent extends LitElement {
```
Add `index.ts` which exports component and styles. The same suggestion for `picking`

tarashaliuk commented on libs/domain/picking/src/components/picking-product-card/picking-product-card.component.ts:
```diff
@@ -0,0 +1,184 @@
+
+  @state() isCorrectNumberOfPickedProvided = true;
+  @state() currentNumberOfPicked?: number;
+  @state() pickedDataEvent?: ProductItemPickedEvent;
```
Looks that this state variable is not used anywhere

tarashaliuk commented on libs/domain/picking/src/components/picking-product-card/picking-product-card.component.ts:
```diff
@@ -0,0 +1,184 @@
+      this.isCorrectNumberOfPickedProvided =
+        0 <= this.currentNumberOfPicked &&
+        this.currentNumberOfPicked <= this.productItem.quantity;
+    }
```
Can we move this logic to `onChangeQuantity` method?

tarashaliuk commented on libs/domain/picking/src/components/picking-product-card/picking-product-card.component.ts:
```diff
@@ -0,0 +1,184 @@
+                  </div>
+
+                  <oryx-button>
+                    <button
+                      ?disabled="${!this.isCorrectNumberOfPickedProvided}"
+                    >
```
`type="submit"`

Sinucid replied:
button without type eq `type=submit`

tarashaliuk commented on libs/domain/picking/src/components/picking-product-card/picking-product-card.component.ts:
```diff
@@ -0,0 +1,184 @@
+        () => html`
+          <oryx-card>
+            <oryx-heading slot="heading" class="title">
+              ${this.productItem?.orderItem?.name}
```
let's use `if (!this.productItem) return` before returning the template to avoid conditional chains

nsilva-spryker replied:
Let's use 'when' directive instead

tarashaliuk replied:
`when` creates one more nested level and requires conditional chains.

nsilva-spryker replied:
Sorry not getting this one, the if is also a conditional, using 'when' I would say it's more lit native since this is essentially a wrapper around a ternary expression. 
We don't see many if statements inside our html templates

tarashaliuk replied:
You can use `if` statement with return because you haven’t `else` state. Benefit of using is better readability of code because you shouldn’t scroll down to see if it is `else` statement and there are less nesting lines. 
Also when you will use `if` statement then you can avoid conditional chains (`this.productItem?.`)

tarashaliuk commented on libs/domain/picking/src/components/picking-product-card/picking-product-card.def.ts:
```diff
@@ -0,0 +1,9 @@
+import { componentDef } from '@spryker-oryx/core';
+
+export const pickingProductCardComponent = componentDef({
+  name: 'oryx-picking-product-card',
+  impl: () =>
+    import('./picking-product-card.component').then(
+      (m) => m.PickingProductCardComponent
+    ),
+});
```
Add schema, please

nsilva-spryker replied:
what's the benefit of using schema ?

tarashaliuk replied:
To be able to use this component in EB

tarashaliuk commented on libs/domain/picking/src/components/picking-product-card/picking-product-card.component.ts:
```diff
@@ -0,0 +1,184 @@
+} from '../../models';
+import { styles } from './picking-product-card.styles';
+
+export class PickingProductCardComponent extends LitElement {
```
Let's have stories for new components

nsilva-spryker replied:
If stories is necessary for this case let's add a sub-task or an independent task fot that, let's not deal with that on this pr

tarashaliuk commented on libs/domain/picking/src/components/picking-product-card/picking-product-card.styles.ts:
```diff
@@ -0,0 +1,70 @@
+    margin: 0 20px;
+  }
+
+  .title {
```
can we use `[slot='heading']` to avoid using extra classes?

tarashaliuk commented on libs/domain/picking/src/components/picking-product-card/picking-product-card.component.ts:
```diff
@@ -0,0 +1,184 @@
+            <oryx-heading slot="heading" class="title">
+              ${this.productItem?.orderItem?.name}
+            </oryx-heading>
+            <div class="subtitle">${this.productItem?.orderItem.sku}</div>
```
we can use `oryx-heading` here and avoid using extra styles

nsilva-spryker replied:
unfortunately not, we need a different style on this one

tarashaliuk replied:
There is `h4` tag with the same styles.
You can find typography styles here:
![image](https://user-images.githubusercontent.com/43376934/228216341-1624206c-56cd-400d-ab0f-105d25931a95.png)

tarashaliuk replied:
Looks that you should update your branch with development

nsilva-spryker replied:
yup, It was updated with this changes from 7 days ago, I will use this instead

tarashaliuk commented on libs/domain/picking/src/components/picking-product-card/picking-product-card.styles.ts:
```diff
@@ -0,0 +1,70 @@
+    color: var(--oryx-color-neutral-400);
+  }
+
+  .edit-quantity {
```
```suggestion
  form {
```

tarashaliuk commented on libs/domain/picking/src/components/picking-product-card/picking-product-card.styles.ts:
```diff
@@ -0,0 +1,70 @@
+    margin-block-start: 7px;
+  }
+
+  .edit-quantity-info {
```
```suggestion
  form div {
```

tarashaliuk commented on libs/domain/picking/src/components/picking/picking.component.ts:
```diff
@@ -56,29 +145,117 @@ export class PickingComponent extends PickingListMixin(LitElement) {
+
+        <span>
+          You only picked
+          <span class="bold-text">
```
```suggestion
          <span class="picked-items-info">
```

tarashaliuk commented on libs/domain/picking/src/components/picking/picking.component.ts:
```diff
@@ -56,29 +145,117 @@ export class PickingComponent extends PickingListMixin(LitElement) {
+          items. Do you really want to complete the pick?
+        </span>
+
+        <div slot="footer">
+          <oryx-button outline type="secondary">
+            <button @click=${this.onModalClose}>
+              ${i18n('picking.product-card.cancel')}
+            </button>
+          </oryx-button>
+
+          <oryx-button type="primary">
+            <button @click=${this.confirmPartialPicking}>
+              ${i18n('picking.product-card.confirm')}
+            </button>
+          </oryx-button>
+        </div>
```
```suggestion
          <oryx-button slot="footer" outline type="secondary">
            <button @click=${this.onModalClose}>
              ${i18n('picking.product-card.cancel')}
            </button>
          </oryx-button>

          <oryx-button slot="footer" type="primary">
            <button @click=${this.confirmPartialPicking}>
              ${i18n('picking.product-card.confirm')}
            </button>
          </oryx-button>
```

tarashaliuk commented on libs/domain/picking/src/components/picking/picking.styles.ts:
```diff
@@ -0,0 +1,42 @@
+    font-weight: 600;
+  }
+
+  oryx-modal [slot='footer'] {
+    width: 100%;
+    display: flex;
+    gap: 10px;
+  }
+
+  oryx-modal oryx-button {
+    width: 100%;
+  }
+`;
```
let's use `footerButtonFullWidth` prop for `oryx-modal` when [Alvina's PR](https://github.com/spryker/oryx/pull/179) will be merged

aleespryker replied:
It's merged, you can use it now.

nsilva-spryker committed:
feature: refactor some code, suggestion from pr

nsilva-spryker committed:
feature: refactor some code, suggestions from pr

nsilva-spryker committed:
Merge branch 'development' into feature/hrz-2411-implement-picking-logic

nsilva-spryker committed:
Merge branch 'development' into feature/hrz-2411-implement-picking-logic

nsilva-spryker committed:
feature: refactor modal with new attribute

Sinucid commented on libs/domain/picking/src/components/picking-product-card/picking-product-card.component.ts:
```diff
@@ -0,0 +1,213 @@
+    }
+  }
+
+  protected getSummaryInfo(): SummaryInfo {
```
Left over I guess?

tarashaliuk commented on libs/domain/picking/src/components/picking-product-card/picking-product-card.component.ts:
```diff
@@ -0,0 +1,213 @@
+            )}
+
+            <oryx-image
+              .src="${this.productItem?.product?.image}"
+              alt="${ifDefined(this.productItem?.orderItem?.name)}"
```
`product` and `orderItem` are not optional

Sinucid commented on libs/domain/picking/src/components/picking/picking.component.ts:
```diff
@@ -1,53 +1,134 @@
+  static styles = styles;
+
+  @state()
+  protected isConfirmPickingDialogOpen = false;
```
Left over? As I suggested before you can just use `partialPicking` as a trigger for confirmation modal

Sinucid commented on libs/domain/picking/src/components/picking/picking.component.ts:
```diff
@@ -1,53 +1,134 @@
+    this.updatePickingItem(productIndex, detail.numberOfPicked);
+  }
+
+  protected editPickingItem(event: Event): void {
```
```
  protected savePickingItem(event: CustomEvent<ProductItemPickedEvent>): void {
    const { productId } = event.detail;
```

Sinucid replied:
I don't see this changes

Sinucid commented on libs/domain/picking/src/components/picking/picking.component.ts:
```diff
@@ -1,53 +1,134 @@
+    ];
+  }
+
+  protected savePickingItem(event: Event): void {
```
```
  protected savePickingItem(event: CustomEvent<ProductItemPickedEvent>): void {
    const { productId, numberOfPicked } = event.detail;
```

Sinucid replied:
Either

nsilva-spryker committed:
Merge branch 'development' into feature/hrz-2411-implement-picking-logic

# Conflicts:
#	libs/domain/picking/src/mocks/src/mock-picking-list.ts

nsilva-spryker committed:
feature: reset local branch with lost commit

nsilva-spryker committed:
feature: refactor some code

nsilva-spryker committed:
feature: refactor some code
